// brs/lib/brsTypes/BrsType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RoString_1 = require("./components/RoString");
/** Set of values supported in BrightScript. */
var ValueKind;
(function (ValueKind) {
    ValueKind[ValueKind["Invalid"] = 0] = "Invalid";
    ValueKind[ValueKind["Boolean"] = 1] = "Boolean";
    ValueKind[ValueKind["String"] = 2] = "String";
    ValueKind[ValueKind["Int32"] = 3] = "Int32";
    ValueKind[ValueKind["Int64"] = 4] = "Int64";
    ValueKind[ValueKind["Float"] = 5] = "Float";
    ValueKind[ValueKind["Double"] = 6] = "Double";
    ValueKind[ValueKind["Callable"] = 7] = "Callable";
    ValueKind[ValueKind["Uninitialized"] = 8] = "Uninitialized";
    ValueKind[ValueKind["Dynamic"] = 9] = "Dynamic";
    ValueKind[ValueKind["Void"] = 10] = "Void";
    ValueKind[ValueKind["Object"] = 11] = "Object";
})(ValueKind = exports.ValueKind || (exports.ValueKind = {}));
(function (ValueKind) {
    /**
     * Converts a `ValueKind` enum member to a human-readable string representation.
     * @returns a textual representation of the provided value kind.
     */
    function toString(kind) {
        switch (kind) {
            case ValueKind.Invalid:
                return "Invalid";
            case ValueKind.Boolean:
                return "Boolean";
            case ValueKind.String:
                return "String";
            case ValueKind.Int32:
                return "Integer";
            case ValueKind.Int64:
                return "LongInteger";
            case ValueKind.Float:
                return "Float";
            case ValueKind.Double:
                return "Double";
            case ValueKind.Callable:
                return "Function";
            case ValueKind.Dynamic:
                return "Dynamic";
            case ValueKind.Void:
                return "Void";
            case ValueKind.Uninitialized:
                return "<UNINITIALIZED>";
            case ValueKind.Object:
                return "Object";
        }
    }
    ValueKind.toString = toString;
    /**
     * Fetches a `ValueKind` enum member by its string representation.
     * @param kind the string representation of a `ValueKind`
     * @returns the corresponding `ValueKind` if one exists, otherwise `undefined`.
     */
    function fromString(kind) {
        switch (kind.toLowerCase()) {
            case "invalid":
                return ValueKind.Invalid;
            case "boolean":
                return ValueKind.Boolean;
            case "string":
                return ValueKind.String;
            case "integer":
                return ValueKind.Int32;
            case "longinteger":
                return ValueKind.Int64;
            case "float":
                return ValueKind.Float;
            case "double":
                return ValueKind.Double;
            case "function":
                return ValueKind.Callable;
            case "dynamic":
                return ValueKind.Dynamic;
            case "void":
                return ValueKind.Void;
            case "<uninitialized>":
                return ValueKind.Uninitialized;
            case "object":
                return ValueKind.Object;
            default:
                return undefined;
        }
    }
    ValueKind.fromString = fromString;
})(ValueKind = exports.ValueKind || (exports.ValueKind = {}));
/** Internal representation of a string in BrightScript. */
class BrsString {
    constructor(value) {
        this.value = value;
        this.kind = ValueKind.String;
    }
    lessThan(other) {
        if (other.kind === ValueKind.String) {
            return BrsBoolean.from(this.value < other.value);
        }
        else if (other instanceof RoString_1.RoString) {
            return this.lessThan(other.unbox());
        }
        return BrsBoolean.False;
    }
    greaterThan(other) {
        if (other.kind === ValueKind.String) {
            return BrsBoolean.from(this.value > other.value);
        }
        else if (other instanceof RoString_1.RoString) {
            return this.greaterThan(other.unbox());
        }
        return BrsBoolean.False;
    }
    equalTo(other) {
        if (other.kind === ValueKind.String) {
            return BrsBoolean.from(this.value === other.value);
        }
        else if (other instanceof RoString_1.RoString) {
            return this.equalTo(other.unbox());
        }
        return BrsBoolean.False;
    }
    toString(parent) {
        return this.value;
    }
    concat(other) {
        return new BrsString(this.value + other.value);
    }
    box() {
        return new RoString_1.RoString(this);
    }
}
exports.BrsString = BrsString;
/** Internal representation of a boolean in BrightScript. */
class BrsBoolean {
    constructor(value) {
        this.value = value;
        this.kind = ValueKind.Boolean;
    }
    toBoolean() {
        return this.value;
    }
    static from(value) {
        return value ? BrsBoolean.True : BrsBoolean.False;
    }
    lessThan(other) {
        // booleans aren't less than anything
        // TODO: Validate on a Roku
        return BrsBoolean.False;
    }
    greaterThan(other) {
        // but isn't greater than anything either
        // TODO: Validate on a Roku
        return BrsBoolean.False;
    }
    equalTo(other) {
        if (other.kind === ValueKind.Boolean) {
            return BrsBoolean.from(this === other);
        }
        return BrsBoolean.False;
    }
    toString(parent) {
        return this.value.toString();
    }
    /**
     * Returns the boolean AND of this value with another value.
     * @param other the other value to AND with this one.
     * @returns `BrsBoolean.True` if both this value and the other are true, otherwise
     *          `BrsBoolean.False`.
     */
    and(other) {
        return BrsBoolean.from(this.value && other.value);
    }
    /**
     * Returns the boolean OR of this value with another value.
     * @param other the other value to AND with this one.
     * @returns `BrsBoolean.True` if either this value or the other are true, otherwise
     *          `BrsBoolean.False`.
     */
    or(other) {
        return BrsBoolean.from(this.value || other.value);
    }
    /**
     * Returns the boolean negation of this value with another value.
     * @returns `BrsBoolean.True` if either this value is false, otherwise
     *          `BrsBoolean.False`.
     */
    not() {
        return BrsBoolean.from(!this.value);
    }
}
BrsBoolean.False = new BrsBoolean(false);
BrsBoolean.True = new BrsBoolean(true);
exports.BrsBoolean = BrsBoolean;
/** Internal representation of the BrightScript `invalid` value. */
class BrsInvalid {
    constructor() {
        this.kind = ValueKind.Invalid;
    }
    lessThan(other) {
        // invalid isn't less than anything
        // TODO: Validate on a Roku
        return BrsBoolean.False;
    }
    greaterThan(other) {
        // but isn't greater than anything either
        // TODO: Validate on a Roku
        return BrsBoolean.False;
    }
    equalTo(other) {
        if (other.kind === ValueKind.Invalid) {
            return BrsBoolean.True;
        }
        return BrsBoolean.False;
    }
    toString(parent) {
        return "invalid";
    }
}
BrsInvalid.Instance = new BrsInvalid();
exports.BrsInvalid = BrsInvalid;
/** Internal representation of uninitialized BrightScript variables. */
class Uninitialized {
    constructor() {
        this.kind = ValueKind.Uninitialized;
    }
    lessThan(other) {
        // uninitialized values aren't less than anything
        return BrsBoolean.False;
    }
    greaterThan(other) {
        // uninitialized values aren't less than anything
        return BrsBoolean.False;
    }
    equalTo(other) {
        if (other.kind === ValueKind.String) {
            // Allow variables to be compared to the string "<UNINITIALIZED>" to test if they've
            // been initialized
            return BrsBoolean.from(other.value === this.toString());
        }
        return BrsBoolean.False;
    }
    toString(parent) {
        return "<UNINITIALIZED>";
    }
}
Uninitialized.Instance = new Uninitialized();
exports.Uninitialized = Uninitialized;
