// brs/lib/brsTypes/Callable.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const Brs = __importStar(require("."));
const Expr = __importStar(require("../parser/Expression"));
const Environment_1 = require("../interpreter/Environment");
/**
 * A variant of the `Argument` interface intended for use only when creating standard library
 * functions.
 */
class StdlibArgument {
    /**
     * Creates an `Argument` without requiring locations to be specified.
     * @param name the name of the argument, used for presentation to users
     * @param type the type of value accepted at runtime
     * @param defaultValue the optional default value to use for this argument if one isn't
     *                     provided at runtime
     */
    constructor(name, type, defaultValue) {
        this.location = StdlibArgument.InternalLocation;
        this.name = { text: name, location: StdlibArgument.InternalLocation };
        this.type = { kind: type, location: StdlibArgument.InternalLocation };
        if (defaultValue) {
            this.defaultValue = new Expr.Literal(defaultValue, StdlibArgument.InternalLocation);
        }
    }
}
/** A fake location exists only within the BRS runtime. */
StdlibArgument.InternalLocation = {
    file: "(stdlib)",
    start: { line: -1, column: -1 },
    end: { line: -1, column: -1 },
};
exports.StdlibArgument = StdlibArgument;
/** The set of possible reasons why a signature and runtime arguments don't match. */
var MismatchReason;
(function (MismatchReason) {
    /** Not enough arguments were provided to satisfy a signature. */
    MismatchReason[MismatchReason["TooFewArguments"] = 0] = "TooFewArguments";
    /** Too many arguments were provided to satisfy a signature. */
    MismatchReason[MismatchReason["TooManyArguments"] = 1] = "TooManyArguments";
    /** An argument's type didn't match the signature's type. */
    MismatchReason[MismatchReason["ArgumentTypeMismatch"] = 2] = "ArgumentTypeMismatch";
})(MismatchReason = exports.MismatchReason || (exports.MismatchReason = {}));
/** A `function` or `sub` (either "native" or implemented in BrightScript) that can be called in a BrightScript file. */
class Callable {
    /**
     * Creates a new BrightScript `function` or `sub`.
     * @param name the name this callable should have within the BrightScript runtime.
     * @param signatures the signatures and associated (JavaScript) implementations this callable should
     *                   have within the BrightScript runtime.
     */
    constructor(name, ...signatures) {
        this.kind = Brs.ValueKind.Callable;
        this.name = name;
        this.signatures = signatures;
    }
    /**
     * Calls the function this `Callable` represents with the provided `arg`uments using the
     * provided `Interpreter` instance.
     *
     * @param interpreter the interpreter to execute this callable in.
     * @param args the arguments to pass to the callable routine.
     *
     * @returns the return value of the function, or `invalid` if nothing is explicitly returned.
     */
    call(interpreter, ...args) {
        let satisfiedSignature = this.getFirstSatisfiedSignature(args);
        if (satisfiedSignature == null) {
            throw new Error("BrightScript function called without first checking for satisfied signatures. " +
                "Ensure `Callable#getAllSignatureMismatches` is called before `Callable#call`.");
        }
        let { signature, impl } = satisfiedSignature;
        let mutableArgs = args.slice();
        return interpreter.inSubEnv(subInterpreter => {
            // first, we need to evaluate all of the parameter default values
            // and define them in a new environment
            signature.args.forEach((param, index) => {
                if (param.defaultValue && mutableArgs[index] == null) {
                    mutableArgs[index] = subInterpreter.evaluate(param.defaultValue);
                }
                subInterpreter.environment.define(Environment_1.Scope.Function, param.name.text, mutableArgs[index]);
            });
            // then return whatever the selected implementation would return
            return impl(subInterpreter, ...mutableArgs);
        });
    }
    lessThan(other) {
        return Brs.BrsBoolean.False;
    }
    greaterThan(other) {
        return Brs.BrsBoolean.False;
    }
    equalTo(other) {
        return Brs.BrsBoolean.from(this === other);
    }
    toString(parent) {
        if (this.name) {
            return `[Function ${this.name}]`;
        }
        else {
            return "[anonymous function]";
        }
    }
    getName() {
        return this.name || "";
    }
    getFirstSatisfiedSignature(args) {
        return this.signatures.filter(sigAndImpl => this.getSignatureMismatches(sigAndImpl.signature, args).length === 0)[0];
    }
    getAllSignatureMismatches(args) {
        return this.signatures.map(sigAndImpl => ({
            signature: sigAndImpl.signature,
            mismatches: this.getSignatureMismatches(sigAndImpl.signature, args),
        }));
    }
    getSignatureMismatches(sig, args) {
        let reasons = [];
        let requiredArgCount = sig.args.filter(arg => !arg.defaultValue).length;
        if (args.length < requiredArgCount) {
            reasons.push({
                reason: MismatchReason.TooFewArguments,
                expected: sig.args.length.toString(),
                received: args.length.toString(),
            });
        }
        else if (args.length > sig.args.length) {
            reasons.push({
                reason: MismatchReason.TooManyArguments,
                expected: sig.args.length.toString(),
                received: args.length.toString(),
            });
        }
        sig.args.slice(0, Math.min(sig.args.length, args.length)).forEach((_value, index) => {
            let expected = sig.args[index];
            let received = args[index];
            if (expected.type.kind === Brs.ValueKind.Dynamic ||
                expected.type.kind === Brs.ValueKind.Object) {
                return;
            }
            if (expected.type.kind !== received.kind) {
                reasons.push({
                    reason: MismatchReason.ArgumentTypeMismatch,
                    expected: Brs.ValueKind.toString(expected.type.kind),
                    received: Brs.ValueKind.toString(received.kind),
                    argName: expected.name.text,
                });
            }
        });
        return reasons;
    }
}
exports.Callable = Callable;
