// brs/lib/brsTypes/Double.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require("./");
const BrsType_1 = require("./BrsType");
const Int32_1 = require("./Int32");
const Int64_1 = require("./Int64");
const Float_1 = require("./Float");
class Double {
    /**
     * Creates a new BrightScript double-precision value representing the provided `value`.
     * @param value the value to store in the BrightScript double, rounded to 64-bit (double)
     *              precision.
     */
    constructor(value) {
        this.kind = BrsType_1.ValueKind.Double;
        this.value = value;
    }
    getValue() {
        return this.value;
    }
    /**
     * Creates a new BrightScript double-precision value representing the floating point value
     * contained in `asString`.
     * @param asString the string representation of the value to store in the BrightScript double.
     *                 Will be rounded to 64-bit (double) precision.
     * @returns a BrightScript double value representing `asString`.
     */
    static fromString(asString) {
        return new Double(Number.parseFloat(asString));
    }
    add(rhs) {
        switch (rhs.kind) {
            case BrsType_1.ValueKind.Int64:
                // TODO: Confirm that (double) + (int64) -> (double)
                return new Double(this.getValue() + rhs.getValue().toNumber());
            case BrsType_1.ValueKind.Int32:
            case BrsType_1.ValueKind.Float:
            case BrsType_1.ValueKind.Double:
                return new Double(this.getValue() + rhs.getValue());
        }
    }
    subtract(rhs) {
        switch (rhs.kind) {
            case BrsType_1.ValueKind.Int64:
                // TODO: Confirm that (double) - (int64) -> (double)
                return new Double(this.getValue() - rhs.getValue().toNumber());
            case BrsType_1.ValueKind.Int32:
            case BrsType_1.ValueKind.Float:
            case BrsType_1.ValueKind.Double:
                return new Double(this.getValue() - rhs.getValue());
        }
    }
    multiply(rhs) {
        switch (rhs.kind) {
            case BrsType_1.ValueKind.Int64:
                // TODO: Confirm that (double) - (int64) -> (double)
                return new Double(this.getValue() * rhs.getValue().toNumber());
            case BrsType_1.ValueKind.Int32:
            case BrsType_1.ValueKind.Float:
            case BrsType_1.ValueKind.Double:
                return new Double(this.getValue() * rhs.getValue());
        }
    }
    divide(rhs) {
        switch (rhs.kind) {
            case BrsType_1.ValueKind.Int64:
                // TODO: Confirm that (double) - (int64) -> (double)
                return new Double(this.getValue() / rhs.getValue().toNumber());
            case BrsType_1.ValueKind.Int32:
            case BrsType_1.ValueKind.Float:
            case BrsType_1.ValueKind.Double:
                return new Double(this.getValue() / rhs.getValue());
        }
    }
    modulo(rhs) {
        switch (rhs.kind) {
            case BrsType_1.ValueKind.Int32:
            case BrsType_1.ValueKind.Float:
            case BrsType_1.ValueKind.Double:
                return new Double(this.getValue() % rhs.getValue());
            case BrsType_1.ValueKind.Int64:
                return new Double(this.getValue() % rhs.getValue().toNumber());
        }
    }
    intDivide(rhs) {
        switch (rhs.kind) {
            case BrsType_1.ValueKind.Int64:
                return new Int64_1.Int64(Math.trunc(this.getValue() / rhs.getValue().toNumber()));
            case BrsType_1.ValueKind.Int32:
            case BrsType_1.ValueKind.Float:
            case BrsType_1.ValueKind.Double:
                return new Int32_1.Int32(Math.trunc(this.getValue() / rhs.getValue()));
        }
    }
    pow(exponent) {
        switch (exponent.kind) {
            case BrsType_1.ValueKind.Int32:
                return new Float_1.Float(Math.pow(this.getValue(), exponent.getValue()));
            case BrsType_1.ValueKind.Int64:
                return new Int64_1.Int64(this.getValue()).pow(exponent);
            case BrsType_1.ValueKind.Float:
                return new Float_1.Float(Math.pow(this.getValue(), exponent.getValue()));
            case BrsType_1.ValueKind.Double:
                return new Double(Math.pow(this.getValue(), exponent.getValue()));
        }
    }
    and(rhs) {
        switch (rhs.kind) {
            case BrsType_1.ValueKind.Int64:
                return new Int64_1.Int64(this.getValue()).and(rhs);
            case BrsType_1.ValueKind.Int32:
            case BrsType_1.ValueKind.Float:
            case BrsType_1.ValueKind.Double:
                return new Int32_1.Int32(this.getValue() & rhs.getValue());
        }
    }
    or(rhs) {
        switch (rhs.kind) {
            case BrsType_1.ValueKind.Int64:
                return new Int64_1.Int64(this.getValue()).or(rhs);
            case BrsType_1.ValueKind.Int32:
            case BrsType_1.ValueKind.Float:
            case BrsType_1.ValueKind.Double:
                return new Int32_1.Int32(this.getValue() | rhs.getValue());
        }
    }
    lessThan(other) {
        switch (other.kind) {
            case BrsType_1.ValueKind.Int64:
                return _1.BrsBoolean.from(this.getValue() < other.getValue().toNumber());
            case BrsType_1.ValueKind.Int32:
            case BrsType_1.ValueKind.Float:
            case BrsType_1.ValueKind.Double:
                return _1.BrsBoolean.from(this.getValue() < other.getValue());
            default:
                return _1.BrsBoolean.False;
        }
    }
    greaterThan(other) {
        switch (other.kind) {
            case BrsType_1.ValueKind.Int64:
                return _1.BrsBoolean.from(this.getValue() > other.getValue().toNumber());
            case BrsType_1.ValueKind.Int32:
            case BrsType_1.ValueKind.Float:
            case BrsType_1.ValueKind.Double:
                return _1.BrsBoolean.from(this.getValue() > other.getValue());
            default:
                return _1.BrsBoolean.False;
        }
    }
    equalTo(other) {
        switch (other.kind) {
            case BrsType_1.ValueKind.Int64:
                return _1.BrsBoolean.from(this.getValue() === other.getValue().toNumber());
            case BrsType_1.ValueKind.Int32:
            case BrsType_1.ValueKind.Float:
            case BrsType_1.ValueKind.Double:
                return _1.BrsBoolean.from(this.getValue() === other.getValue());
            default:
                return _1.BrsBoolean.False;
        }
    }
    toString(parent) {
        return this.value.toString();
    }
}
exports.Double = Double;
