// brs/lib/stdlib/Json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RoAssociativeArray_1 = require("../brsTypes/components/RoAssociativeArray");
const RoArray_1 = require("../brsTypes/components/RoArray");
const brsTypes_1 = require("../brsTypes");
/**
 * Converts a value to its representation as a BrsType. If no such
 * representation is possible, throws an Error.
 * @param {any} x Some value.
 * @return {BrsType} The BrsType representaion of `x`.
 * @throws {Error} If `x` cannot be represented as a BrsType.
 */
function brsValueOf(x) {
    if (x === null) {
        return brsTypes_1.BrsInvalid.Instance;
    }
    let t = typeof x;
    switch (t) {
        case "boolean":
            return brsTypes_1.BrsBoolean.from(x);
        case "string":
            return new brsTypes_1.BrsString(x);
        case "number":
            if (Number.isInteger(x)) {
                return x >= -2147483648 && x <= 2147483647 ? new brsTypes_1.Int32(x) : new brsTypes_1.Int64(x);
            }
            return new brsTypes_1.Float(x);
        case "object":
            if (Array.isArray(x)) {
                return new RoArray_1.RoArray(x.map(brsValueOf));
            }
            return new RoAssociativeArray_1.RoAssociativeArray(Object.getOwnPropertyNames(x).map((k) => ({
                name: new brsTypes_1.BrsString(k),
                value: brsValueOf(x[k]),
            })));
        default:
            throw new Error(`brsValueOf not implemented for: ${x} <${t}>`);
    }
}
function visit(x, visited) {
    if (visited.has(x)) {
        throw new Error("Nested object reference");
    }
    visited.add(x);
}
/**
 * Converts a BrsType value to its representation as a JSON string. If no such
 * representation is possible, throws an Error. Objects with cyclical references
 * are rejected.
 * @param {Interpreter} interpreter An Interpreter.
 * @param {BrsType} x Some BrsType value.
 * @param {Set<BrsAggregate>} visited An optional Set of visited of RoArray or
 *   RoAssociativeArray. If not provided, a new Set will be created.
 * @return {string} The JSON string representation of `x`.
 * @throws {Error} If `x` cannot be represented as a JSON string.
 */
function jsonOf(interpreter, x, visited = new Set()) {
    switch (x.kind) {
        case brsTypes_1.ValueKind.Invalid:
            return "null";
        case brsTypes_1.ValueKind.String:
            return `"${x.toString()}"`;
        case brsTypes_1.ValueKind.Boolean:
        case brsTypes_1.ValueKind.Double:
        case brsTypes_1.ValueKind.Float:
        case brsTypes_1.ValueKind.Int32:
        case brsTypes_1.ValueKind.Int64:
            return x.toString();
        case brsTypes_1.ValueKind.Object:
            if (x instanceof RoAssociativeArray_1.RoAssociativeArray) {
                visit(x, visited);
                return `{${x
                    .getElements()
                    .map((k) => {
                    return `"${k.toString()}":${jsonOf(interpreter, x.get(k), visited)}`;
                })
                    .join(",")}}`;
            }
            if (x instanceof RoArray_1.RoArray) {
                visit(x, visited);
                return `[${x
                    .getElements()
                    .map((el) => {
                    return jsonOf(interpreter, el, visited);
                })
                    .join(",")}]`;
            }
            break;
        case brsTypes_1.ValueKind.Callable:
        case brsTypes_1.ValueKind.Uninitialized:
            break;
        default:
            // Exhaustive check as per:
            // https://basarat.gitbooks.io/typescript/content/docs/types/discriminated-unions.html
            const _ = x;
            break;
    }
    throw new Error(`jsonOf not implemented for: ${x}`);
}
function logBrsErr(functionName, err) {
    if (process.env.NODE_ENV === "test") {
        return;
    }
    console.error(`BRIGHTSCRIPT: ERROR: ${functionName}: ${err.message}`);
}
exports.FormatJson = new brsTypes_1.Callable("FormatJson", {
    signature: {
        returns: brsTypes_1.ValueKind.String,
        args: [
            new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Object),
            new brsTypes_1.StdlibArgument("flags", brsTypes_1.ValueKind.Int32, new brsTypes_1.Int32(0)),
        ],
    },
    impl: (interpreter, x, _flags) => {
        try {
            return new brsTypes_1.BrsString(jsonOf(interpreter, x));
        }
        catch (err) {
            // example RBI error:
            // "BRIGHTSCRIPT: ERROR: FormatJSON: Value type not supported: roFunction: pkg:/source/main.brs(14)"
            logBrsErr("FormatJSON", err);
            return new brsTypes_1.BrsString("");
        }
    },
});
exports.ParseJson = new brsTypes_1.Callable("ParseJson", {
    signature: {
        returns: brsTypes_1.ValueKind.Dynamic,
        args: [new brsTypes_1.StdlibArgument("jsonString", brsTypes_1.ValueKind.String)],
    },
    impl: (_, jsonString) => {
        try {
            let s = jsonString.toString().trim();
            if (s === "") {
                throw new Error("Data is empty");
            }
            return brsValueOf(JSON.parse(s));
        }
        catch (err) {
            // example RBI error:
            // "BRIGHTSCRIPT: ERROR: ParseJSON: Unknown identifier 'x': pkg:/source/main.brs(25)"
            logBrsErr("ParseJSON", err);
            return brsTypes_1.BrsInvalid.Instance;
        }
    },
});
