// brs/lib/stdlib/Math.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const brsTypes_1 = require("../brsTypes");
/** Returns the absolute value of a float. */
exports.Abs = new brsTypes_1.Callable("Abs", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Float)],
        returns: brsTypes_1.ValueKind.Float,
    },
    impl: (interpreter, x) => new brsTypes_1.Float(Math.abs(x.getValue())),
});
/*
 * Returns the integer as a 32-bit float.
 * ** NOTE: the function name implies it makes a 64-bit float, but the docs say
 *     it currently returns a 32-bit float, but may return a 64-bit float in the future.
 */
exports.Cdbl = new brsTypes_1.Callable("Cdbl", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Int32)],
        returns: brsTypes_1.ValueKind.Float,
    },
    impl: (interpreter, x) => new brsTypes_1.Float(x.getValue()),
});
/** Returns an integer from a float rounding up from midpoints */
exports.Cint = new brsTypes_1.Callable("Cint", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Float)],
        returns: brsTypes_1.ValueKind.Int32,
    },
    impl: (interpreter, x) => new brsTypes_1.Int32(Math.round(x.getValue())),
});
/** Returns the integer as a 32-bit float. */
exports.Csng = new brsTypes_1.Callable("Csng", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Int32)],
        returns: brsTypes_1.ValueKind.Float,
    },
    impl: (interpreter, x) => new brsTypes_1.Float(x.getValue()),
});
/** Returns an integer from a float removing fractional parts. */
exports.Fix = new brsTypes_1.Callable("Fix", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Float)],
        returns: brsTypes_1.ValueKind.Int32,
    },
    impl: (interpreter, x) => new brsTypes_1.Int32(Math.trunc(x.getValue())),
});
/** Returns an integer from a float. */
exports.Int = new brsTypes_1.Callable("Int", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Float)],
        returns: brsTypes_1.ValueKind.Int32,
    },
    impl: (interpreter, x) => new brsTypes_1.Int32(Math.floor(x.getValue())),
});
function SgnImpl(interpreter, x) {
    let val = x.getValue();
    if (val > 0.0)
        return new brsTypes_1.Int32(1);
    else if (val < 0.0)
        return new brsTypes_1.Int32(-1);
    else
        return new brsTypes_1.Int32(0);
}
/** Returns -1 if parameter is negative, 0 if zero, and 1 if positive. */
exports.Sgn = new brsTypes_1.Callable("Sgn", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Float)],
        returns: brsTypes_1.ValueKind.Int32,
    },
    impl: SgnImpl,
}, {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Int32)],
        returns: brsTypes_1.ValueKind.Int32,
    },
    impl: SgnImpl,
});
/** Returns the arc-tangent (in radians) of a float. */
exports.Atn = new brsTypes_1.Callable("Atn", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Float)],
        returns: brsTypes_1.ValueKind.Float,
    },
    impl: (interpreter, x) => new brsTypes_1.Float(Math.atan(x.getValue())),
});
/** Returns the cosine of a float (argument must be provided in radians). */
exports.Cos = new brsTypes_1.Callable("Cos", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Float)],
        returns: brsTypes_1.ValueKind.Float,
    },
    impl: (interpreter, x) => new brsTypes_1.Float(Math.cos(x.getValue())),
});
/** Returns the sine of a float (argument must be provided in radians). */
exports.Sin = new brsTypes_1.Callable("Sin", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Float)],
        returns: brsTypes_1.ValueKind.Float,
    },
    impl: (interpreter, x) => new brsTypes_1.Float(Math.sin(x.getValue())),
});
/** Returns the tangent float (argument must be provided in radians). */
exports.Tan = new brsTypes_1.Callable("Tan", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Float)],
        returns: brsTypes_1.ValueKind.Float,
    },
    impl: (interpreter, x) => new brsTypes_1.Float(Math.tan(x.getValue())),
});
/** Returns the natural exponent of a float. */
exports.Exp = new brsTypes_1.Callable("Exp", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Float)],
        returns: brsTypes_1.ValueKind.Float,
    },
    impl: (interpreter, x) => new brsTypes_1.Float(Math.exp(x.getValue())),
});
/** Returns the log of a float. */
exports.Log = new brsTypes_1.Callable("Log", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Float)],
        returns: brsTypes_1.ValueKind.Float,
    },
    impl: (interpreter, x) => new brsTypes_1.Float(Math.log(x.getValue())),
});
/** Returns the square root of a float. */
exports.Sqr = new brsTypes_1.Callable("Sqr", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("x", brsTypes_1.ValueKind.Float)],
        returns: brsTypes_1.ValueKind.Float,
    },
    impl: (interpreter, x) => new brsTypes_1.Float(Math.sqrt(x.getValue())),
});
/**
 * Returns a random number in a given range. If the range is zero, a random
 * float between [0,1) is returned. If the range is a positive number, a
 * random integer between 1 and that number is returned (inclusive is returned).
 *
 * **NOTE:** the float returned is in the range [0,1) to match the javascript
 *     implementation, while the brightscript specification calls for (0,1).
 *     This should be okay in practice, but if this is necessary a more complicated
 *     implementation will be necessary.
 */
exports.Rnd = new brsTypes_1.Callable("Rnd", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("range", brsTypes_1.ValueKind.Int32)],
        returns: brsTypes_1.ValueKind.Dynamic,
    },
    impl: (interpreter, range) => {
        if (range.getValue() === 0)
            return new brsTypes_1.Float(Math.random());
        else
            return new brsTypes_1.Int32(Math.floor(Math.random() * range.getValue() + 1));
    },
});
