// brs/lib/stdlib/Print.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const brsTypes_1 = require("../brsTypes");
/**
 * Moves the cursor to the specified position on the current line. If the
 * provided position is greater than the current console width and the output
 * is a TTY, the resulting position is modulo'd by the current console width.
 * May be used several times in a `print` list.
 */
exports.Tab = new brsTypes_1.Callable("Tab", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("position", brsTypes_1.ValueKind.Int32)],
        returns: brsTypes_1.ValueKind.String,
    },
    impl: (interpreter, position) => {
        let target = position.getValue();
        if (target < 0 || target < interpreter.stdout.position()) {
            return new brsTypes_1.BrsString("");
        }
        // TODO: this probably won't handle text wrapping well, but I'm not
        // sure what the reference implementation does here yet
        return new brsTypes_1.BrsString(" ".repeat(target - interpreter.stdout.position()));
    },
});
/**
 * Returns a number from 0 to the current console width, indicating the
 * position of the output cursor. Requires a "dummy argument" of any type, as
 * it's completely ignored.
 */
exports.Pos = new brsTypes_1.Callable("Pos", {
    signature: {
        // `pos` expects an argument and doesn't use it. The reference
        // implementation's documentation even says it must be provided but
        // isn't used: https://sdkdocs.roku.com/display/sdkdoc/Program+Statements#ProgramStatements-PRINTitemlist
        args: [new brsTypes_1.StdlibArgument("dummy", brsTypes_1.ValueKind.Dynamic)],
        returns: brsTypes_1.ValueKind.Int32,
    },
    impl: (interpreter) => {
        return new brsTypes_1.Int32(interpreter.stdout.position());
    },
});
