// brs/lib/stdlib/Run.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = __importStar(require("path"));
const brs = __importStar(require("../"));
const brsTypes_1 = require("../brsTypes");
const File_1 = require("./File");
/**
 * Runs a file (or set of files) with the provided arguments, returning either the value returned by those files'
 * `main` function or `invalid` if an error occurs.
 *
 * @param interpreter the interpreter hosting this call to `Run`
 * @param filenames a list of files to lex, parse, and run
 * @param args the arguments to pass into the found `main` function
 *
 * @returns the value returned by the executed file(s) if no errors are detected, otherwise `invalid`
 */
function runFiles(interpreter, filenames, args) {
    let volumes = filenames.map(filename => File_1.getVolumeByPath(interpreter, filename.value));
    let pathsToFiles = filenames.map(filename => path.join(interpreter.options.root, File_1.getPath(filename.value)));
    // if the file-to-run doesn't exist, RBI returns invalid
    if (!volumes.every(volume => volume != null)) {
        return brsTypes_1.BrsInvalid.Instance;
    }
    try {
        let results = brs.executeSync(pathsToFiles, interpreter.options, args);
        return results[0] || brsTypes_1.BrsInvalid.Instance;
    }
    catch (err) {
        // swallow errors and just return invalid; RBI returns invalid for "file doesn't exist" errors,
        // syntax errors, etc.
        return brsTypes_1.BrsInvalid.Instance;
    }
}
/**
 * Creates several copies of the provided signature and implementation pair, simulating variadic types by creating a
 * function that accepts zero args, one that accepts one arg, one that accepts two args, (…).
 *
 * @param signatureAndImpl the base signature and implementation to make variadic
 *
 * @returns an array containing psuedo-variadic versions of the provided signature and implementation
 */
function variadic(signatureAndImpl) {
    let { signature, impl } = signatureAndImpl;
    return [
        signatureAndImpl,
        ...new Array(10).fill(0).map((_, numArgs) => {
            return {
                signature: {
                    args: [
                        ...signature.args,
                        ...new Array(numArgs)
                            .fill(0)
                            .map((_, i) => new brsTypes_1.StdlibArgument(`arg${i}`, brsTypes_1.ValueKind.Dynamic)),
                    ],
                    returns: signature.returns,
                },
                impl: impl,
            };
        }),
    ];
}
exports.Run = new brsTypes_1.Callable("Run", ...variadic({
    signature: {
        args: [new brsTypes_1.StdlibArgument("filename", brsTypes_1.ValueKind.String)],
        returns: brsTypes_1.ValueKind.Dynamic,
    },
    impl: (interpreter, filename, ...args) => {
        return runFiles(interpreter, [filename], args);
    },
}), ...variadic({
    signature: {
        args: [new brsTypes_1.StdlibArgument("filenamearray", brsTypes_1.ValueKind.Object)],
        returns: brsTypes_1.ValueKind.Dynamic,
    },
    impl: (interpreter, filenamearray, ...args) => {
        if (filenamearray instanceof brsTypes_1.RoArray &&
            filenamearray.getElements().every(brsTypes_1.isBrsString)) {
            return runFiles(interpreter, filenamearray.getElements(), args);
        }
        // RBI seems to hard-reboot when passed a non-empty associative array, but returns invalid for empty
        // AA's. Let's return invalid to be safe.
        return brsTypes_1.BrsInvalid.Instance;
    },
}));
