// brs/lib/stdlib/String.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const brsTypes_1 = require("../brsTypes");
/** Converts the string to all uppercase. */
exports.UCase = new brsTypes_1.Callable("UCase", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("s", brsTypes_1.ValueKind.String)],
        returns: brsTypes_1.ValueKind.String,
    },
    impl: (interpreter, s) => new brsTypes_1.BrsString(s.value.toUpperCase()),
});
/** Converts the string to all lowercase. */
exports.LCase = new brsTypes_1.Callable("LCase", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("s", brsTypes_1.ValueKind.String)],
        returns: brsTypes_1.ValueKind.String,
    },
    impl: (interpreter, s) => new brsTypes_1.BrsString(s.value.toLowerCase()),
});
/**
 * Returns the Unicode ("ASCII") value for the first character of the specified string.
 * An empty string argument will return `0`.
 */
exports.Asc = new brsTypes_1.Callable("Asc", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("letter", brsTypes_1.ValueKind.String)],
        returns: brsTypes_1.ValueKind.String,
    },
    impl: (interpreter, str) => new brsTypes_1.Int32(str.value.charCodeAt(0) || 0),
});
/**
 * Performs the inverse of the `Asc` function: returns a one-character string whose character has
 * the specified Unicode value.
 *
 * Returns empty string (`""`) if the specified value is `0` or an invalid Unicode value.
 */
exports.Chr = new brsTypes_1.Callable("Chr", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("ch", brsTypes_1.ValueKind.Int32)],
        returns: brsTypes_1.ValueKind.String,
    },
    impl: (interpreter, ch) => {
        const num = ch.getValue();
        if (num <= 0)
            return new brsTypes_1.BrsString("");
        else
            return new brsTypes_1.BrsString(String.fromCharCode(num));
    },
});
/**
 * Returns the first n characters in a string
 */
exports.Left = new brsTypes_1.Callable("Left", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("s", brsTypes_1.ValueKind.String), new brsTypes_1.StdlibArgument("n", brsTypes_1.ValueKind.Int32)],
        returns: brsTypes_1.ValueKind.String,
    },
    impl: (interpreter, s, n) => new brsTypes_1.BrsString(s.value.substr(0, n.getValue())),
});
/**
 * Returns the last n characters in a string
 */
exports.Right = new brsTypes_1.Callable("Right", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("s", brsTypes_1.ValueKind.String), new brsTypes_1.StdlibArgument("n", brsTypes_1.ValueKind.Int32)],
        returns: brsTypes_1.ValueKind.String,
    },
    impl: (interpreter, s, n) => {
        let end = s.value.length - 1;
        let start = end - (n.getValue() - 1);
        if (n.getValue() <= 0)
            return new brsTypes_1.BrsString("");
        else if (start < 0)
            return new brsTypes_1.BrsString(s.value);
        return new brsTypes_1.BrsString(s.value.substr(start, end));
    },
});
/**
 * Returns the index (1 based) of a string inside another string. Zero if there is no match.
 */
exports.Instr = new brsTypes_1.Callable("Instr", {
    signature: {
        args: [
            new brsTypes_1.StdlibArgument("start", brsTypes_1.ValueKind.Int32),
            new brsTypes_1.StdlibArgument("str", brsTypes_1.ValueKind.String),
            new brsTypes_1.StdlibArgument("search", brsTypes_1.ValueKind.String),
        ],
        returns: brsTypes_1.ValueKind.String,
    },
    impl: (interpreter, start, str, search) => new brsTypes_1.Int32(str.value.indexOf(search.value, start.getValue() - 1) + 1),
});
/**
 * Return the number of characters in a string
 */
exports.Len = new brsTypes_1.Callable("Len", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("s", brsTypes_1.ValueKind.String)],
        returns: brsTypes_1.ValueKind.Int32,
    },
    impl: (interpreter, s) => new brsTypes_1.Int32(s.value.length),
});
/**
 * Return a string located in the middle of another string from start index to end index
 */
exports.Mid = new brsTypes_1.Callable("Mid", {
    signature: {
        args: [
            new brsTypes_1.StdlibArgument("s", brsTypes_1.ValueKind.String),
            new brsTypes_1.StdlibArgument("p", brsTypes_1.ValueKind.Int32),
        ],
        returns: brsTypes_1.ValueKind.String,
    },
    impl: (interpreter, s, p) => {
        let start = p.getValue() - 1;
        return new brsTypes_1.BrsString(s.value.substring(start));
    },
}, {
    signature: {
        args: [
            new brsTypes_1.StdlibArgument("s", brsTypes_1.ValueKind.String),
            new brsTypes_1.StdlibArgument("p", brsTypes_1.ValueKind.Int32),
            new brsTypes_1.StdlibArgument("n", brsTypes_1.ValueKind.Int32),
        ],
        returns: brsTypes_1.ValueKind.String,
    },
    impl: (interpreter, s, p, n) => {
        let start = p.getValue() - 1;
        return new brsTypes_1.BrsString(s.value.substring(start, start + n.getValue()));
    },
});
/**
 * Return a string from a float. If it is positive, prefix it with a space.
 */
exports.Str = new brsTypes_1.Callable("Str", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("value", brsTypes_1.ValueKind.Float)],
        returns: brsTypes_1.ValueKind.String,
    },
    impl: (interpreter, value) => {
        const floatValue = value.getValue();
        const prefix = floatValue > 0.0 ? " " : "";
        return new brsTypes_1.BrsString(prefix + String(floatValue));
    },
});
/**
 * Return a string from an integer. If it is positive, prefix it with a space.
 */
exports.StrI = new brsTypes_1.Callable("StrI", {
    signature: {
        args: [
            new brsTypes_1.StdlibArgument("value", brsTypes_1.ValueKind.Int32),
            new brsTypes_1.StdlibArgument("radix", brsTypes_1.ValueKind.Int32, new brsTypes_1.Int32(10)),
        ],
        returns: brsTypes_1.ValueKind.String,
    },
    impl: (interpreter, value, brsRadix) => {
        let radix = brsRadix.getValue();
        if (radix < 2 || radix > 36) {
            return new brsTypes_1.BrsString("");
        }
        const intValue = value.getValue();
        const prefix = intValue > 0 && radix === 10 ? " " : "";
        return new brsTypes_1.BrsString(prefix + intValue.toString(radix));
    },
});
/**
 * Return a string from another string replacing instances of {index} with the
 * respective parameter.
 */
exports.Substitute = new brsTypes_1.Callable("Substitute", {
    signature: {
        args: [
            new brsTypes_1.StdlibArgument("str", brsTypes_1.ValueKind.String),
            new brsTypes_1.StdlibArgument("arg0", brsTypes_1.ValueKind.String),
            new brsTypes_1.StdlibArgument("arg1", brsTypes_1.ValueKind.String, new brsTypes_1.BrsString("")),
            new brsTypes_1.StdlibArgument("arg2", brsTypes_1.ValueKind.String, new brsTypes_1.BrsString("")),
            new brsTypes_1.StdlibArgument("arg3", brsTypes_1.ValueKind.String, new brsTypes_1.BrsString("")),
        ],
        returns: brsTypes_1.ValueKind.String,
    },
    impl: (interpreter, str, arg0, arg1, arg2, arg3) => {
        let completelyReplaced = [arg0, arg1, arg2, arg3].reduce((replaced, replacement, index) => replaced.replace(new RegExp(`\\{${index}\\}`, "g"), replacement.value), str.value);
        return new brsTypes_1.BrsString(completelyReplaced);
    },
});
/**
 * Return a float or integer from a string
 */
exports.Val = new brsTypes_1.Callable("Val", {
    signature: {
        args: [
            new brsTypes_1.StdlibArgument("s", brsTypes_1.ValueKind.String),
            new brsTypes_1.StdlibArgument("radix", brsTypes_1.ValueKind.Int32, new brsTypes_1.Int32(10)),
        ],
        returns: brsTypes_1.ValueKind.Dynamic,
    },
    impl: (interpreter, s, brsRadix) => {
        function isBrsStrFloat(str) {
            return str.value.includes(".");
        }
        if (isBrsStrFloat(s)) {
            return new brsTypes_1.Float(Number(s.value));
        }
        else {
            return new brsTypes_1.Int32(parseInt(s.value, brsRadix.getValue()));
        }
    },
});
/**
 * Return an integer from a string or 0 if it can't be parsed.
 */
exports.StrToI = new brsTypes_1.Callable("StrToI", {
    signature: {
        args: [new brsTypes_1.StdlibArgument("s", brsTypes_1.ValueKind.String)],
        returns: brsTypes_1.ValueKind.Int32,
    },
    impl: (interpreter, s) => {
        let integerValue = parseInt(s.value);
        if (Number.isNaN(integerValue)) {
            return new brsTypes_1.Int32(0);
        }
        return new brsTypes_1.Int32(integerValue);
    },
});
