// brs/lib/interpreter/AstPrinter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** Creates a pretty-printed representation of an expression to ease debugging. */
class AstPrinter {
    constructor() {
        this.indent = 0;
    }
    /**
     * Pretty-prints an expression for debugging purposes.
     * @param expression the expression to pretty-print.
     */
    print(expression) {
        this.indent = 0;
        return expression.accept(this);
    }
    visitAnonymousFunction(e) {
        return JSON.stringify(e, undefined, 2);
    }
    visitBinary(e) {
        return this.parenthesize(e.token.text, e.left, e.right);
    }
    visitCall(e) {
        return JSON.stringify(e, undefined, 2);
    }
    visitDottedGet(e) {
        return JSON.stringify(e, undefined, 2);
    }
    visitIndexedGet(e) {
        return JSON.stringify(e, undefined, 2);
    }
    visitGrouping(e) {
        return this.parenthesize("group", e.expression);
    }
    visitLiteral(e) {
        if (e.value == null) {
            return "invalid";
        }
        else {
            return e.value.toString();
        }
    }
    visitArrayLiteral(e) {
        return JSON.stringify(e, undefined, 2);
    }
    visitAALiteral(e) {
        return JSON.stringify(e, undefined, 2);
    }
    visitDottedSet(e) {
        return JSON.stringify(e, undefined, 2);
    }
    visitIndexedSet(e) {
        return JSON.stringify(e, undefined, 2);
    }
    visitIncrement(e) {
        return JSON.stringify(e, undefined, 2);
    }
    visitUnary(e) {
        return this.parenthesize(e.operator.text, e.right);
    }
    visitVariable(expression) {
        return JSON.stringify(expression, undefined, 2);
    }
    /**
     * Wraps an expression in parentheses to make its grouping visible during debugging.
     *
     * @param name The name of the expression type being printed.
     * @param expressions any subexpressions that need to be stringified as well.
     */
    parenthesize(name = "", ...expressions) {
        this.indent++;
        let out = [
            `(${name}\n`,
            expressions.map(e => `${"  ".repeat(this.indent)}${e.accept(this)}\n`).join(""),
            `${"  ".repeat(this.indent - 1)})`,
        ].join("");
        this.indent--;
        return out;
    }
}
exports.AstPrinter = AstPrinter;
