// rooibos-cli/dist/lib/Feedback.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FileFeedback_1 = require("./FileFeedback");
var _feedback = [];
function feedbackMessage(type, file, message, isThrown) {
    if (isThrown === void 0) { isThrown = false; }
    var error = new FileFeedback_1.FileFeedback(file, type, message);
    _feedback.push(error);
    if (isThrown) {
        error.throw();
    }
}
exports.feedbackMessage = feedbackMessage;
function resetFeedback() {
    _feedback.splice(0, _feedback.length);
}
exports.resetFeedback = resetFeedback;
function feedbackInfo(file, message, isThrown) {
    if (isThrown === void 0) { isThrown = false; }
    feedbackMessage(FileFeedback_1.FileFeedbackType.Info, file, message, isThrown);
}
exports.feedbackInfo = feedbackInfo;
function feedbackWarning(file, message, isThrown) {
    if (isThrown === void 0) { isThrown = false; }
    feedbackMessage(FileFeedback_1.FileFeedbackType.Warning, file, message, isThrown);
}
exports.feedbackWarning = feedbackWarning;
function feedbackError(file, message, isThrown) {
    if (isThrown === void 0) { isThrown = false; }
    feedbackMessage(FileFeedback_1.FileFeedbackType.Error, file, message, isThrown);
}
exports.feedbackError = feedbackError;
function getFeedbackErrors() {
    return _feedback.filter(function (feedback) { return feedback.feedbackType === FileFeedback_1.FileFeedbackType.Error; });
}
exports.getFeedbackErrors = getFeedbackErrors;
function getFeedbackWarnings() {
    return _feedback.filter(function (feedback) { return feedback.feedbackType === FileFeedback_1.FileFeedbackType.Warning; });
}
exports.getFeedbackWarnings = getFeedbackWarnings;
function getFeedbackInfos() {
    return _feedback.filter(function (feedback) { return feedback.feedbackType === FileFeedback_1.FileFeedbackType.Info; });
}
exports.getFeedbackInfos = getFeedbackInfos;
