// rooibos-cli/dist/lib/FileFeedback.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FileFeedbackType;
(function (FileFeedbackType) {
    FileFeedbackType["Info"] = "Info";
    FileFeedbackType["Warning"] = "Warning";
    FileFeedbackType["Error"] = "Error";
})(FileFeedbackType = exports.FileFeedbackType || (exports.FileFeedbackType = {}));
var FileFeedback = /** @class */ (function () {
    function FileFeedback(file, feedbackType, message, line, character) {
        this.file = file;
        this.feedbackType = feedbackType;
        this.message = message;
        this.line = line;
        this.character = character;
    }
    FileFeedback.prototype.toString = function () {
        var fileName = this.file ? this.file.fullPath : 'No file';
        var lineText = this.line ? this.line.toString() : '-';
        var charText = this.character ? this.character.toString() : '-';
        return this.feedbackType.toString() + " - " + fileName + "(" + lineText + ":" + charText + ") " + this.message;
    };
    FileFeedback.prototype.throw = function () {
        throw new Error(this.toString());
    };
    return FileFeedback;
}());
exports.FileFeedback = FileFeedback;
