// brs/lib/lexer/Lexeme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Lexeme;
(function (Lexeme) {
    // parens (and friends)
    Lexeme["LeftParen"] = "LeftParen";
    Lexeme["RightParen"] = "RightParen";
    Lexeme["LeftSquare"] = "LeftSquare";
    Lexeme["RightSquare"] = "RightSquare";
    Lexeme["LeftBrace"] = "LeftBrace";
    Lexeme["RightBrace"] = "RightBrace";
    // operators
    Lexeme["Caret"] = "Caret";
    Lexeme["Minus"] = "Minus";
    Lexeme["Plus"] = "Plus";
    Lexeme["Star"] = "Star";
    Lexeme["Slash"] = "Slash";
    Lexeme["Mod"] = "Mod";
    Lexeme["Backslash"] = "Backslash";
    // postfix operators
    Lexeme["PlusPlus"] = "PlusPlus";
    Lexeme["MinusMinus"] = "MinusMinus";
    // bitshift
    Lexeme["LeftShift"] = "LeftShift";
    Lexeme["RightShift"] = "RightShift";
    // assignment operators
    Lexeme["MinusEqual"] = "MinusEqual";
    Lexeme["PlusEqual"] = "PlusEqual";
    Lexeme["StarEqual"] = "StarEqual";
    Lexeme["SlashEqual"] = "SlashEqual";
    Lexeme["BackslashEqual"] = "BackslashEqual";
    Lexeme["LeftShiftEqual"] = "LeftShiftEqual";
    Lexeme["RightShiftEqual"] = "RightShiftEqual";
    // comparators
    Lexeme["Less"] = "Less";
    Lexeme["LessEqual"] = "LessEqual";
    Lexeme["Greater"] = "Greater";
    Lexeme["GreaterEqual"] = "GreaterEqual";
    Lexeme["Equal"] = "Equal";
    Lexeme["LessGreater"] = "LessGreater";
    // literals
    Lexeme["Identifier"] = "Identifier";
    Lexeme["String"] = "String";
    Lexeme["Integer"] = "Integer";
    Lexeme["Float"] = "Float";
    Lexeme["Double"] = "Double";
    Lexeme["LongInteger"] = "LongInteger";
    // other single-character symbols
    Lexeme["Dot"] = "Dot";
    Lexeme["Comma"] = "Comma";
    Lexeme["Colon"] = "Colon";
    Lexeme["Semicolon"] = "Semicolon";
    // conditional compilation
    Lexeme["HashIf"] = "HashIf";
    Lexeme["HashElseIf"] = "HashElseIf";
    Lexeme["HashElse"] = "HashElse";
    Lexeme["HashEndIf"] = "HashEndIf";
    Lexeme["HashConst"] = "HashConst";
    Lexeme["HashError"] = "HashError";
    Lexeme["HashErrorMessage"] = "HashErrorMessage";
    // keywords
    // canonical source: https://sdkdocs.roku.com/display/sdkdoc/Reserved+Words
    Lexeme["And"] = "And";
    Lexeme["Box"] = "Box";
    Lexeme["CreateObject"] = "CreateObject";
    Lexeme["Dim"] = "Dim";
    Lexeme["Else"] = "Else";
    Lexeme["ElseIf"] = "ElseIf";
    Lexeme["End"] = "End";
    Lexeme["EndFunction"] = "EndFunction";
    Lexeme["EndFor"] = "EndFor";
    Lexeme["EndIf"] = "EndIf";
    Lexeme["EndSub"] = "EndSub";
    Lexeme["EndWhile"] = "EndWhile";
    Lexeme["Eval"] = "Eval";
    Lexeme["Exit"] = "Exit";
    Lexeme["ExitFor"] = "ExitFor";
    Lexeme["ExitWhile"] = "ExitWhile";
    Lexeme["False"] = "False";
    Lexeme["For"] = "For";
    Lexeme["ForEach"] = "ForEach";
    Lexeme["Function"] = "Function";
    Lexeme["GetGlobalAA"] = "GetGlobalAA";
    Lexeme["GetLastRunCompileError"] = "GetLastRunCompileError";
    Lexeme["GetLastRunRunTimeError"] = "GetLastRunRunTimeError";
    Lexeme["Goto"] = "Goto";
    Lexeme["If"] = "If";
    Lexeme["Invalid"] = "Invalid";
    Lexeme["Let"] = "Let";
    Lexeme["Next"] = "Next";
    Lexeme["Not"] = "Not";
    Lexeme["ObjFun"] = "ObjFun";
    Lexeme["Or"] = "Or";
    Lexeme["Pos"] = "Pos";
    Lexeme["Print"] = "Print";
    Lexeme["Rem"] = "Rem";
    Lexeme["Return"] = "Return";
    Lexeme["Step"] = "Step";
    Lexeme["Stop"] = "Stop";
    Lexeme["Sub"] = "Sub";
    Lexeme["Tab"] = "Tab";
    Lexeme["To"] = "To";
    Lexeme["True"] = "True";
    Lexeme["Type"] = "Type";
    Lexeme["While"] = "While";
    // structural
    Lexeme["Newline"] = "Newline";
    Lexeme["Eof"] = "Eof";
})(Lexeme = exports.Lexeme || (exports.Lexeme = {}));
