// brs/lib/lexer/ReservedWords.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Lexeme_1 = require("./Lexeme");
/**
 * The set of all reserved words in the reference BrightScript runtime. These can't be used for any
 * other purpose within a BrightScript file.
 * @see https://sdkdocs.roku.com/display/sdkdoc/Reserved+Words
 */
exports.ReservedWords = new Set([
    "and",
    "box",
    "createobject",
    "dim",
    "each",
    "else",
    "elseif",
    "endsub",
    "endwhile",
    "eval",
    "exit",
    "exitwhile",
    "false",
    "for",
    "function",
    "getglobalaa",
    "getlastruncompileerror",
    "getlastrunruntimeerror",
    "goto",
    "if",
    "invalid",
    "let",
    "line_num",
    "next",
    "not",
    "objfun",
    "or",
    "pos",
    "print",
    "rem",
    "return",
    "run",
    "step",
    "stop",
    "sub",
    "tab",
    "then",
    "to",
    "true",
    "type",
    "while",
]);
/**
 * The set of keywords in the reference BrightScript runtime. Any of these that *are not* reserved
 * words can be used within a BrightScript file for other purposes, e.g. `tab`.
 *
 * Unfortunately there's no canonical source for this!
 */
exports.KeyWords = {
    and: Lexeme_1.Lexeme.And,
    dim: Lexeme_1.Lexeme.Dim,
    else: Lexeme_1.Lexeme.Else,
    elseif: Lexeme_1.Lexeme.ElseIf,
    "else if": Lexeme_1.Lexeme.ElseIf,
    endfor: Lexeme_1.Lexeme.EndFor,
    "end for": Lexeme_1.Lexeme.EndFor,
    endfunction: Lexeme_1.Lexeme.EndFunction,
    "end function": Lexeme_1.Lexeme.EndFunction,
    endif: Lexeme_1.Lexeme.EndIf,
    "end if": Lexeme_1.Lexeme.EndIf,
    endsub: Lexeme_1.Lexeme.EndSub,
    "end sub": Lexeme_1.Lexeme.EndSub,
    endwhile: Lexeme_1.Lexeme.EndWhile,
    "end while": Lexeme_1.Lexeme.EndWhile,
    exit: Lexeme_1.Lexeme.Exit,
    "exit for": Lexeme_1.Lexeme.ExitFor,
    exitwhile: Lexeme_1.Lexeme.ExitWhile,
    "exit while": Lexeme_1.Lexeme.ExitWhile,
    false: Lexeme_1.Lexeme.False,
    for: Lexeme_1.Lexeme.For,
    "for each": Lexeme_1.Lexeme.ForEach,
    function: Lexeme_1.Lexeme.Function,
    goto: Lexeme_1.Lexeme.Goto,
    if: Lexeme_1.Lexeme.If,
    invalid: Lexeme_1.Lexeme.Invalid,
    let: Lexeme_1.Lexeme.Let,
    mod: Lexeme_1.Lexeme.Mod,
    next: Lexeme_1.Lexeme.Next,
    not: Lexeme_1.Lexeme.Not,
    or: Lexeme_1.Lexeme.Or,
    print: Lexeme_1.Lexeme.Print,
    rem: Lexeme_1.Lexeme.Rem,
    return: Lexeme_1.Lexeme.Return,
    step: Lexeme_1.Lexeme.Step,
    stop: Lexeme_1.Lexeme.Stop,
    sub: Lexeme_1.Lexeme.Sub,
    to: Lexeme_1.Lexeme.To,
    true: Lexeme_1.Lexeme.True,
    while: Lexeme_1.Lexeme.While,
};
