// brs/lib/brsTypes/components/RoArray.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const BrsType_1 = require("../BrsType");
const BrsComponent_1 = require("./BrsComponent");
const Callable_1 = require("../Callable");
const Int32_1 = require("../Int32");
class RoArray extends BrsComponent_1.BrsComponent {
    constructor(elements) {
        super("roArray");
        this.kind = BrsType_1.ValueKind.Object;
        this.peek = new Callable_1.Callable("peek", {
            signature: {
                args: [],
                returns: BrsType_1.ValueKind.Dynamic,
            },
            impl: (interpreter) => {
                return this.elements[this.elements.length - 1] || BrsType_1.BrsInvalid.Instance;
            },
        });
        this.pop = new Callable_1.Callable("pop", {
            signature: {
                args: [],
                returns: BrsType_1.ValueKind.Dynamic,
            },
            impl: (interpreter) => {
                return this.elements.pop() || BrsType_1.BrsInvalid.Instance;
            },
        });
        this.push = new Callable_1.Callable("push", {
            signature: {
                args: [new Callable_1.StdlibArgument("talue", BrsType_1.ValueKind.Dynamic)],
                returns: BrsType_1.ValueKind.Void,
            },
            impl: (interpreter, tvalue) => {
                this.elements.push(tvalue);
                return BrsType_1.BrsInvalid.Instance;
            },
        });
        this.shift = new Callable_1.Callable("shift", {
            signature: {
                args: [],
                returns: BrsType_1.ValueKind.Dynamic,
            },
            impl: (interpreter) => {
                return this.elements.shift() || BrsType_1.BrsInvalid.Instance;
            },
        });
        this.unshift = new Callable_1.Callable("unshift", {
            signature: {
                args: [new Callable_1.StdlibArgument("tvalue", BrsType_1.ValueKind.Dynamic)],
                returns: BrsType_1.ValueKind.Void,
            },
            impl: (interpreter, tvalue) => {
                this.elements.unshift(tvalue);
                return BrsType_1.BrsInvalid.Instance;
            },
        });
        this.delete = new Callable_1.Callable("delete", {
            signature: {
                args: [new Callable_1.StdlibArgument("index", BrsType_1.ValueKind.Int32)],
                returns: BrsType_1.ValueKind.Boolean,
            },
            impl: (interpreter, index) => {
                if (index.lessThan(new Int32_1.Int32(0)).toBoolean()) {
                    return BrsType_1.BrsBoolean.False;
                }
                let deleted = this.elements.splice(index.getValue(), 1);
                return BrsType_1.BrsBoolean.from(deleted.length > 0);
            },
        });
        this.count = new Callable_1.Callable("count", {
            signature: {
                args: [],
                returns: BrsType_1.ValueKind.Int32,
            },
            impl: (interpreter) => {
                return new Int32_1.Int32(this.elements.length);
            },
        });
        this.clear = new Callable_1.Callable("clear", {
            signature: {
                args: [],
                returns: BrsType_1.ValueKind.Void,
            },
            impl: (interpreter) => {
                this.elements = [];
                return BrsType_1.BrsInvalid.Instance;
            },
        });
        this.append = new Callable_1.Callable("append", {
            signature: {
                args: [new Callable_1.StdlibArgument("array", BrsType_1.ValueKind.Object)],
                returns: BrsType_1.ValueKind.Void,
            },
            impl: (interpreter, array) => {
                if (!(array instanceof RoArray)) {
                    // TODO: validate against RBI
                    return BrsType_1.BrsInvalid.Instance;
                }
                this.elements = [
                    ...this.elements,
                    ...array.elements.filter(element => !!element),
                ];
                return BrsType_1.BrsInvalid.Instance;
            },
        });
        this.elements = elements;
        this.registerMethods([
            this.peek,
            this.pop,
            this.push,
            this.shift,
            this.unshift,
            this.delete,
            this.count,
            this.clear,
            this.append,
        ]);
    }
    toString(parent) {
        if (parent) {
            return "<Component: roArray>";
        }
        return [
            "<Component: roArray> =",
            "[",
            ...this.elements.map((el) => `    ${el.toString(this)}`),
            "]",
        ].join("\n");
    }
    equalTo(other) {
        return BrsType_1.BrsBoolean.False;
    }
    getValue() {
        return this.elements;
    }
    getElements() {
        return this.elements.slice();
    }
    get(index) {
        switch (index.kind) {
            case BrsType_1.ValueKind.Int32:
                return this.getElements()[index.getValue()] || BrsType_1.BrsInvalid.Instance;
            case BrsType_1.ValueKind.String:
                return this.getMethod(index.value) || BrsType_1.BrsInvalid.Instance;
            default:
                throw new Error("Array indexes must be 32-bit integers, or method names must be strings");
        }
    }
    set(index, value) {
        if (index.kind !== BrsType_1.ValueKind.Int32) {
            throw new Error("Array indexes must be 32-bit integers");
        }
        this.elements[index.getValue()] = value;
        return BrsType_1.BrsInvalid.Instance;
    }
}
exports.RoArray = RoArray;
