// brs/lib/brsTypes/components/RoAssociativeArray.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const BrsType_1 = require("../BrsType");
const BrsComponent_1 = require("./BrsComponent");
const Callable_1 = require("../Callable");
const Int32_1 = require("../Int32");
const RoArray_1 = require("./RoArray");
class RoAssociativeArray extends BrsComponent_1.BrsComponent {
    constructor(elements) {
        super("roAssociativeArray");
        this.kind = BrsType_1.ValueKind.Object;
        this.elements = new Map();
        /** Removes all elements from the associative array */
        this.clear = new Callable_1.Callable("clear", {
            signature: {
                args: [],
                returns: BrsType_1.ValueKind.Void,
            },
            impl: (interpreter) => {
                this.elements.clear();
                return BrsType_1.BrsInvalid.Instance;
            },
        });
        /** Removes a given item from the associative array */
        this.delete = new Callable_1.Callable("delete", {
            signature: {
                args: [new Callable_1.StdlibArgument("str", BrsType_1.ValueKind.String)],
                returns: BrsType_1.ValueKind.Boolean,
            },
            impl: (interpreter, str) => {
                let deleted = this.elements.delete(str.value);
                return BrsType_1.BrsBoolean.from(deleted);
            },
        });
        /** Given a key and value, adds an item to the associative array if it doesn't exist
         * Or replaces the value of a key that already exists in the associative array
         */
        this.addreplace = new Callable_1.Callable("addreplace", {
            signature: {
                args: [
                    new Callable_1.StdlibArgument("key", BrsType_1.ValueKind.String),
                    new Callable_1.StdlibArgument("value", BrsType_1.ValueKind.Dynamic),
                ],
                returns: BrsType_1.ValueKind.Void,
            },
            impl: (interpreter, key, value) => {
                this.set(key, value);
                return BrsType_1.BrsInvalid.Instance;
            },
        });
        /** Returns the number of items in the associative array */
        this.count = new Callable_1.Callable("count", {
            signature: {
                args: [],
                returns: BrsType_1.ValueKind.Int32,
            },
            impl: (interpreter) => {
                return new Int32_1.Int32(this.elements.size);
            },
        });
        /** Returns a boolean indicating whether or not a given key exists in the associative array */
        this.doesexist = new Callable_1.Callable("doesexist", {
            signature: {
                args: [new Callable_1.StdlibArgument("str", BrsType_1.ValueKind.String)],
                returns: BrsType_1.ValueKind.Boolean,
            },
            impl: (interpreter, str) => {
                return this.get(str) !== BrsType_1.BrsInvalid.Instance ? BrsType_1.BrsBoolean.True : BrsType_1.BrsBoolean.False;
            },
        });
        /** Appends a new associative array to another. If two keys are the same, the value of the original AA is replaced with the new one. */
        this.append = new Callable_1.Callable("append", {
            signature: {
                args: [new Callable_1.StdlibArgument("obj", BrsType_1.ValueKind.Object)],
                returns: BrsType_1.ValueKind.Void,
            },
            impl: (interpreter, obj) => {
                if (!(obj instanceof RoAssociativeArray)) {
                    // TODO: validate against RBI
                    return BrsType_1.BrsInvalid.Instance;
                }
                this.elements = new Map([...this.elements, ...obj.elements]);
                return BrsType_1.BrsInvalid.Instance;
            },
        });
        /** Returns an array of keys from the associative array in lexicographical order */
        this.keys = new Callable_1.Callable("keys", {
            signature: {
                args: [],
                returns: BrsType_1.ValueKind.Object,
            },
            impl: (interpreter) => {
                return new RoArray_1.RoArray(this.getElements());
            },
        });
        /** Returns an array of values from the associative array in lexicographical order */
        this.items = new Callable_1.Callable("items", {
            signature: {
                args: [],
                returns: BrsType_1.ValueKind.Object,
            },
            impl: (interpreter) => {
                return new RoArray_1.RoArray(this.getValues());
            },
        });
        /** Given a key, returns the value associated with that key. This method is case insensitive. */
        this.lookup = new Callable_1.Callable("lookup", {
            signature: {
                args: [new Callable_1.StdlibArgument("key", BrsType_1.ValueKind.String)],
                returns: BrsType_1.ValueKind.Dynamic,
            },
            impl: (interpreter, key) => {
                let lKey = key.value.toLowerCase();
                return this.get(new BrsType_1.BrsString(lKey));
            },
        });
        elements.forEach(member => this.elements.set(member.name.value.toLowerCase(), member.value));
        this.registerMethods([
            this.clear,
            this.delete,
            this.addreplace,
            this.count,
            this.doesexist,
            this.append,
            this.keys,
            this.items,
            this.lookup,
        ]);
    }
    toString(parent) {
        if (parent) {
            return "<Component: roAssociativeArray>";
        }
        return [
            "<Component: roAssociativeArray> =",
            "{",
            ...Array.from(this.elements.keys()).map(key => `    ${key}: ${this.elements.get(key).toString(this)}`),
            "}",
        ].join("\n");
    }
    equalTo(other) {
        return BrsType_1.BrsBoolean.False;
    }
    getValue() {
        return this.elements;
    }
    getElements() {
        return Array.from(this.elements.keys())
            .sort()
            .map(key => new BrsType_1.BrsString(key));
    }
    getValues() {
        return Array.from(this.elements.values())
            .sort()
            .map((value) => value);
    }
    get(index) {
        if (index.kind !== BrsType_1.ValueKind.String) {
            throw new Error("Associative array indexes must be strings");
        }
        // TODO: this works for now, in that a property with the same name as a method essentially
        // overwrites the method. The only reason this doesn't work is that getting a method from an
        // associative array and _not_ calling it returns `invalid`, but calling it returns the
        // function itself. I'm not entirely sure why yet, but it's gotta have something to do with
        // how methods are implemented within RBI.
        //
        // Are they stored separately from elements, like they are here? Or does
        // `Interpreter#visitCall` need to check for `invalid` in its callable, then try to find a
        // method with the desired name separately? That last bit would work but it's pretty gross.
        // That'd allow roArrays to have methods with the methods not accessible via `arr["count"]`.
        // Same with RoAssociativeArrays I guess.
        return (this.elements.get(index.value.toLowerCase()) ||
            this.getMethod(index.value) ||
            BrsType_1.BrsInvalid.Instance);
    }
    set(index, value) {
        if (index.kind !== BrsType_1.ValueKind.String) {
            throw new Error("Associative array indexes must be strings");
        }
        this.elements.set(index.value.toLowerCase(), value);
        return BrsType_1.BrsInvalid.Instance;
    }
}
exports.RoAssociativeArray = RoAssociativeArray;
