// brs/lib/brsTypes/components/RoRegex.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const BrsType_1 = require("../BrsType");
const BrsComponent_1 = require("./BrsComponent");
const Callable_1 = require("../Callable");
const RoArray_1 = require("./RoArray");
class RoRegex extends BrsComponent_1.BrsComponent {
    constructor(expression, flags = new BrsType_1.BrsString("")) {
        super("roRegex");
        this.kind = BrsType_1.ValueKind.Object;
        this.supportedFlags = "ims";
        /** Returns whether the string matched the regex or not */
        this.isMatch = new Callable_1.Callable("ismatch", {
            signature: {
                args: [new Callable_1.StdlibArgument("str", BrsType_1.ValueKind.String)],
                returns: BrsType_1.ValueKind.Boolean,
            },
            impl: (interpreter, str) => {
                return BrsType_1.BrsBoolean.from(this.jsRegex.test(str.value));
            },
        });
        /** Returns an array of matches */
        this.match = new Callable_1.Callable("match", {
            signature: {
                args: [new Callable_1.StdlibArgument("str", BrsType_1.ValueKind.String)],
                returns: BrsType_1.ValueKind.Object,
            },
            impl: (interpreter, str) => {
                const result = this.jsRegex.exec(str.value);
                let arr = [];
                if (result !== null) {
                    arr = result.map(match => new BrsType_1.BrsString(match || ""));
                }
                return new RoArray_1.RoArray(arr);
            },
        });
        /** Returns a new string with first match replaced */
        this.replace = new Callable_1.Callable("replace", {
            signature: {
                args: [
                    new Callable_1.StdlibArgument("str", BrsType_1.ValueKind.String),
                    new Callable_1.StdlibArgument("replacement", BrsType_1.ValueKind.String),
                ],
                returns: BrsType_1.ValueKind.String,
            },
            impl: (interpreter, str, replacement) => {
                let replacementPattern = this.parseReplacementPattern(replacement.value);
                const newStr = this.jsRegex[Symbol.replace](str.value, replacementPattern);
                return new BrsType_1.BrsString(newStr);
            },
        });
        /** Returns a new string with all matches replaced */
        this.replaceAll = new Callable_1.Callable("replaceall", {
            signature: {
                args: [
                    new Callable_1.StdlibArgument("str", BrsType_1.ValueKind.String),
                    new Callable_1.StdlibArgument("replacement", BrsType_1.ValueKind.String),
                ],
                returns: BrsType_1.ValueKind.String,
            },
            impl: (interpreter, str, replacement) => {
                const source = this.jsRegex.source;
                const flags = this.jsRegex.flags + "g";
                this.jsRegex = new RegExp(source, flags);
                const newStr = this.jsRegex[Symbol.replace](str.value, replacement.value);
                return new BrsType_1.BrsString(newStr);
            },
        });
        /** Returns an array of strings split by match */
        this.split = new Callable_1.Callable("split", {
            signature: {
                args: [new Callable_1.StdlibArgument("str", BrsType_1.ValueKind.String)],
                returns: BrsType_1.ValueKind.Object,
            },
            impl: (interpreter, str) => {
                let items = this.jsRegex[Symbol.split](str.value);
                let brsItems = items.map(item => new BrsType_1.BrsString(item));
                return new RoArray_1.RoArray(brsItems);
            },
        });
        /** Returns an array of array with all matches found */
        this.matchAll = new Callable_1.Callable("matchall", {
            signature: {
                args: [new Callable_1.StdlibArgument("str", BrsType_1.ValueKind.String)],
                returns: BrsType_1.ValueKind.Object,
            },
            impl: (interpreter, str) => {
                const source = this.jsRegex.source;
                const flags = this.jsRegex.flags + "g";
                this.jsRegex = new RegExp(source, flags);
                let arr = [];
                let matches;
                while ((matches = this.jsRegex.exec(str.value)) !== null) {
                    let item = new BrsType_1.BrsString(matches[0] || "");
                    arr.push(new RoArray_1.RoArray([item]));
                }
                return new RoArray_1.RoArray(arr);
            },
        });
        this.jsRegex = new RegExp(expression.value, this.parseFlags(flags.value));
        this.registerMethods([
            this.isMatch,
            this.match,
            this.replace,
            this.replaceAll,
            this.split,
            this.matchAll,
        ]);
    }
    toString(parent) {
        return "<Component: roRegex>";
    }
    equalTo(other) {
        return BrsType_1.BrsBoolean.False;
    }
    /**
     * Checks and parses the flags to avoid passing flags
     * that are not supported
     * @param inputFlags Flags passed to constructor
     * @returns parsed flags
     */
    parseFlags(inputFlags) {
        let parsedFlags = "";
        if (inputFlags.length === 0) {
            return parsedFlags;
        }
        for (const flag of inputFlags) {
            if (flag === "x") {
                console.warn("'x' flag is not implemented yet, ignoring flag.");
            }
            else if (!this.supportedFlags.includes(flag)) {
                throw new Error(`${flag} is not supported.`);
            }
            else {
                parsedFlags += flag;
            }
        }
        return parsedFlags;
    }
    /**
     * Transforms positional pattern replacements to javascript syntax
     * by replacing backslashes with dollar symbols
     * @param pattern Pattern to replace
     * @returns Replaced string
     */
    parseReplacementPattern(pattern) {
        return pattern.replace(/\\/g, "$");
    }
}
exports.RoRegex = RoRegex;
