// brs/lib/brsTypes/components/Timespan.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const BrsType_1 = require("../BrsType");
const BrsComponent_1 = require("./BrsComponent");
const Callable_1 = require("../Callable");
const Int32_1 = require("../Int32");
const luxon = __importStar(require("luxon"));
class Timespan extends BrsComponent_1.BrsComponent {
    constructor() {
        super("roTimespan");
        this.kind = BrsType_1.ValueKind.Object;
        this.markTime = Date.now();
        /** Sets timespan object to the current time */
        this.mark = new Callable_1.Callable("mark", {
            signature: {
                args: [],
                returns: BrsType_1.ValueKind.Void,
            },
            impl: (_) => {
                this.resetTime();
                return BrsType_1.BrsInvalid.Instance;
            },
        });
        /** Returns total milliseconds from the mark time to now */
        this.totalmilliseconds = new Callable_1.Callable("totalmilliseconds", {
            signature: {
                args: [],
                returns: BrsType_1.ValueKind.Int32,
            },
            impl: (_) => {
                return new Int32_1.Int32(Date.now() - this.markTime);
            },
        });
        /** Returns total seconds from the mark time to now */
        this.totalseconds = new Callable_1.Callable("totalseconds", {
            signature: {
                args: [],
                returns: BrsType_1.ValueKind.Int32,
            },
            impl: (_) => {
                return new Int32_1.Int32((Date.now() - this.markTime) / 1000);
            },
        });
        /** Parses an ISO8601 date and returns number of seconds from now until the given date.
         * If the date is not a valid ISO8601 date string and can't be parsed, the int 2077252342 is returned, consistent with the brightscript method.
         */
        this.getsecondstoiso8601date = new Callable_1.Callable("getsecondstoiso8601date", {
            signature: {
                args: [new Callable_1.StdlibArgument("date", BrsType_1.ValueKind.String)],
                returns: BrsType_1.ValueKind.Int32,
            },
            impl: (_, date) => {
                let dateAsSeconds;
                let now = Date.now();
                let dateToParse = luxon.DateTime.fromISO(date.value, { zone: "utc" });
                if (dateToParse.isValid) {
                    dateAsSeconds = (Date.parse(dateToParse.toISO()) - now) / 1000;
                }
                else {
                    dateAsSeconds = 2077252342;
                }
                return new Int32_1.Int32(dateAsSeconds);
            },
        });
        this.registerMethods([
            this.mark,
            this.totalmilliseconds,
            this.totalseconds,
            this.getsecondstoiso8601date,
        ]);
        this.resetTime();
    }
    resetTime() {
        this.markTime = Date.now();
    }
    toString(parent) {
        return "<Component: roTimespan>";
    }
    equalTo(other) {
        return BrsType_1.BrsBoolean.False;
    }
}
exports.Timespan = Timespan;
