// brs/lib/parser/BlockEndReason.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Error_1 = require("../Error");
/** Marker class for errors thrown to exit block execution early. */
class BlockEnd extends Error_1.BrsError {
}
exports.BlockEnd = BlockEnd;
/** An error thrown to exit a for loop early. */
class ExitForReason extends BlockEnd {
    constructor(location) {
        super("`exit for` encountered", location);
    }
}
exports.ExitForReason = ExitForReason;
/** An error thrown to exit a while loop early. */
class ExitWhileReason extends BlockEnd {
    constructor(location) {
        super("`exit while` encountered", location);
    }
}
exports.ExitWhileReason = ExitWhileReason;
/** An error thrown to handle a `return` statement. */
class ReturnValue extends BlockEnd {
    constructor(location, value) {
        super("`return` encountered", location);
        this.location = location;
        this.value = value;
    }
}
exports.ReturnValue = ReturnValue;
/** An error thrown when a BrightScript runtime error is encountered. */
class Runtime extends BlockEnd {
}
exports.Runtime = Runtime;
