// brs/lib/parser/Expression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Binary {
    constructor(left, token, right) {
        this.left = left;
        this.token = token;
        this.right = right;
    }
    accept(visitor) {
        return visitor.visitBinary(this);
    }
    get location() {
        return {
            file: this.token.location.file,
            start: this.left.location.start,
            end: this.right.location.end,
        };
    }
}
exports.Binary = Binary;
class Call {
    constructor(callee, closingParen, args) {
        this.callee = callee;
        this.closingParen = closingParen;
        this.args = args;
    }
    accept(visitor) {
        return visitor.visitCall(this);
    }
    get location() {
        return {
            file: this.closingParen.location.file,
            start: this.callee.location.start,
            end: this.closingParen.location.end,
        };
    }
}
Call.MaximumArguments = 32;
exports.Call = Call;
class Function {
    constructor(parameters, returns, body, keyword, end) {
        this.parameters = parameters;
        this.returns = returns;
        this.body = body;
        this.keyword = keyword;
        this.end = end;
    }
    accept(visitor) {
        return visitor.visitAnonymousFunction(this);
    }
    get location() {
        return {
            file: this.keyword.location.file,
            start: this.keyword.location.start,
            end: this.end.location.end,
        };
    }
}
exports.Function = Function;
class DottedGet {
    constructor(obj, name) {
        this.obj = obj;
        this.name = name;
    }
    accept(visitor) {
        return visitor.visitDottedGet(this);
    }
    get location() {
        return {
            file: this.obj.location.file,
            start: this.obj.location.start,
            end: this.name.location.end,
        };
    }
}
exports.DottedGet = DottedGet;
class IndexedGet {
    constructor(obj, index, closingSquare) {
        this.obj = obj;
        this.index = index;
        this.closingSquare = closingSquare;
    }
    accept(visitor) {
        return visitor.visitIndexedGet(this);
    }
    get location() {
        return {
            file: this.obj.location.file,
            start: this.obj.location.start,
            end: this.closingSquare.location.end,
        };
    }
}
exports.IndexedGet = IndexedGet;
class Grouping {
    constructor(tokens, expression) {
        this.tokens = tokens;
        this.expression = expression;
    }
    accept(visitor) {
        return visitor.visitGrouping(this);
    }
    get location() {
        return {
            file: this.tokens.left.location.file,
            start: this.tokens.left.location.start,
            end: this.tokens.right.location.end,
        };
    }
}
exports.Grouping = Grouping;
class Literal {
    constructor(value, _location) {
        this.value = value;
        this._location = _location;
    }
    accept(visitor) {
        return visitor.visitLiteral(this);
    }
    get location() {
        return (this._location || {
            file: "(internal)",
            start: {
                line: -1,
                column: -1,
            },
            end: {
                line: -1,
                column: -1,
            },
        });
    }
}
exports.Literal = Literal;
class ArrayLiteral {
    constructor(elements, open, close) {
        this.elements = elements;
        this.open = open;
        this.close = close;
    }
    accept(visitor) {
        return visitor.visitArrayLiteral(this);
    }
    get location() {
        return {
            file: this.open.location.file,
            start: this.open.location.start,
            end: this.close.location.end,
        };
    }
}
exports.ArrayLiteral = ArrayLiteral;
class AALiteral {
    constructor(elements, open, close) {
        this.elements = elements;
        this.open = open;
        this.close = close;
    }
    accept(visitor) {
        return visitor.visitAALiteral(this);
    }
    get location() {
        return {
            file: this.open.location.file,
            start: this.open.location.start,
            end: this.close.location.end,
        };
    }
}
exports.AALiteral = AALiteral;
class Unary {
    constructor(operator, right) {
        this.operator = operator;
        this.right = right;
    }
    accept(visitor) {
        return visitor.visitUnary(this);
    }
    get location() {
        return {
            file: this.operator.location.file,
            start: this.operator.location.start,
            end: this.right.location.end,
        };
    }
}
exports.Unary = Unary;
class Variable {
    constructor(name) {
        this.name = name;
    }
    accept(visitor) {
        return visitor.visitVariable(this);
    }
    get location() {
        return this.name.location;
    }
}
exports.Variable = Variable;
