// brs/lib/preprocessor/Chunk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** A series of BrightScript tokens that will be parsed and interpreted directly. */
class BrightScript {
    constructor(tokens) {
        this.tokens = tokens;
    }
    accept(visitor) {
        return visitor.visitBrightScript(this);
    }
}
exports.BrightScript = BrightScript;
/**
 * A conditional compilation directive that declares a constant value that's in-scope only during
 * preprocessing.
 *
 * Typically takes the form of:
 *
 * @example
 * #const foo = true
 */
class Declaration {
    constructor(name, value) {
        this.name = name;
        this.value = value;
    }
    accept(visitor) {
        return visitor.visitDeclaration(this);
    }
}
exports.Declaration = Declaration;
/**
 * A directive that adds the "conditional" to "conditional compilation". Typically takes the form
 * of:
 *
 * @example
 * #if foo
 *     someBrightScriptGoesHere()
 * #else if bar
 *     compileSomeOtherCode()
 * #else
 *     otherwise("compile this!")
 * #end if
 */
class If {
    constructor(condition, thenChunks, elseIfs, elseChunks) {
        this.condition = condition;
        this.thenChunks = thenChunks;
        this.elseIfs = elseIfs;
        this.elseChunks = elseChunks;
    }
    accept(visitor) {
        return visitor.visitIf(this);
    }
}
exports.If = If;
/**
 * A forced BrightScript compilation error with a message attached.  Typically takes the form of:
 *
 * @example
 * #error Some message describing the error goes here.
 */
class Error {
    constructor(hashError, message) {
        this.hashError = hashError;
        this.message = message;
    }
    accept(visitor) {
        return visitor.visitError(this);
    }
}
exports.Error = Error;
