// brs/lib/preprocessor/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const Parser_1 = require("./Parser");
const Preprocessor_1 = require("./Preprocessor");
const Manifest_1 = require("./Manifest");
class Preprocessor {
    constructor() {
        this.parser = new Parser_1.Parser();
        this._preprocessor = new Preprocessor_1.Preprocessor();
        this.events = new events_1.EventEmitter();
        // plumb errors from the internal parser and preprocessor out to the public interface for convenience
        this.parser.events.on("err", err => this.events.emit("err", err));
        this._preprocessor.events.on("err", err => this.events.emit("err", err));
    }
    /**
     * Convenience function to subscribe to the `err` events emitted by `preprocessor.events`.
     * @param errorHandler the function to call for every preprocessing error emitted after subscribing
     * @returns an object with a `dispose` function, used to unsubscribe from errors
     */
    onError(errorHandler) {
        this.events.on("err", errorHandler);
        return {
            dispose: () => {
                this.events.removeListener("err", errorHandler);
            },
        };
    }
    /**
     * Convenience function to subscribe to a single `err` event emitted by `preprocessor.events`.
     * @param errorHandler the function to call for the first preprocessing error emitted after subscribing
     */
    onErrorOnce(errorHandler) {
        this.events.once("err", errorHandler);
    }
    /**
     * Pre-processes a set of tokens, evaluating any conditional compilation directives encountered.
     * @param tokens the set of tokens to process
     * @param manifest the data stored in the found manifest file
     * @returns an array of processed tokens representing a subset of the provided ones
     */
    preprocess(tokens, manifest) {
        let parserResults = this.parser.parse(tokens);
        if (parserResults.errors.length > 0) {
            return {
                processedTokens: [],
                errors: parserResults.errors,
            };
        }
        return this._preprocessor.filter(parserResults.chunks, Manifest_1.getBsConst(manifest));
    }
}
exports.Preprocessor = Preprocessor;
const Chunk = __importStar(require("./Chunk"));
exports.Chunk = Chunk;
var Parser_2 = require("./Parser");
exports.Parser = Parser_2.Parser;
var Manifest_2 = require("./Manifest");
exports.getManifest = Manifest_2.getManifest;
exports.getManifestSync = Manifest_2.getManifestSync;
exports.getBsConst = Manifest_2.getBsConst;
