// brs/lib/preprocessor/Parser.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const lexer_1 = require("../lexer");
const parser_1 = require("../parser");
const CC = __importStar(require("./Chunk"));
/** * Parses `Tokens` into chunks of tokens, excluding conditional compilation directives. */
class Parser {
    constructor() {
        this.events = new events_1.EventEmitter();
    }
    /**
     * Parses an array of tokens into an array of "chunks" - conditional compilation directives and their
     * associated BrightScript.
     *
     * @param toParse the array of tokens to parse
     * @returns an array of chunks (conditional compilation directives and the associated BrightScript) to be later
     *          executed.
     */
    parse(toParse) {
        let current = 0;
        let tokens = toParse;
        let errors = [];
        /**
         * Emits an error via this parser's `events` property, then throws it.
         * @param err the ParseError to emit then throw
         */
        const emitError = (err) => {
            errors.push(err);
            this.events.emit("err", err);
            throw err;
        };
        try {
            return {
                chunks: nChunks(),
                errors: errors,
            };
        }
        catch (conditionalCompilationError) {
            return {
                chunks: [],
                errors: errors,
            };
        }
        /**
         * Parses tokens to produce an array containing a variable number of heterogeneous chunks.
         * @returns a heterogeneous array of chunks
         */
        function nChunks() {
            let chunks = [];
            while (true) {
                let c = hashConst();
                if (c) {
                    chunks.push(c);
                }
                let maybeEof = eof();
                if (maybeEof) {
                    chunks.push(maybeEof);
                    break;
                }
                else if (!c) {
                    break;
                }
            }
            return chunks;
        }
        /**
         * Parses tokens to produce a "declaration" chunk if possible, otherwise falls back to `hashIf`.
         * @returns a "declaration" chunk if one is detected, otherwise whatever `hashIf` returns
         */
        function hashConst() {
            if (match(lexer_1.Lexeme.HashConst)) {
                let name = advance();
                consume("Expected '=' after #const (name)", lexer_1.Lexeme.Equal);
                let value = advance();
                match(lexer_1.Lexeme.Newline);
                return new CC.Declaration(name, value);
            }
            return hashIf();
        }
        /**
         * Parses tokens to produce an "if" chunk (including "else if" and "else" chunks) if possible,
         * otherwise falls back to `hashError`.
         * @returns an "if" chunk if one is detected, otherwise whatever `hashError` returns
         */
        function hashIf() {
            if (match(lexer_1.Lexeme.HashIf)) {
                let startingLine = previous().location.start.line;
                let elseChunk;
                let ifCondition = advance();
                match(lexer_1.Lexeme.Newline);
                let thenChunk = nChunks();
                let elseIfs = [];
                while (match(lexer_1.Lexeme.HashElseIf)) {
                    let condition = advance();
                    match(lexer_1.Lexeme.Newline);
                    elseIfs.push({
                        condition: condition,
                        thenChunks: nChunks(),
                    });
                }
                if (match(lexer_1.Lexeme.HashElse)) {
                    match(lexer_1.Lexeme.Newline);
                    elseChunk = nChunks();
                }
                consume(`Expected '#else if' to close '#if' conditional compilation statement starting on line ${startingLine}`, lexer_1.Lexeme.HashEndIf);
                match(lexer_1.Lexeme.Newline);
                return new CC.If(ifCondition, thenChunk, elseIfs, elseChunk);
            }
            return hashError();
        }
        /**
         * Parses tokens to produce an "error" chunk (including the associated message) if possible,
         * otherwise falls back to a chunk of plain BrightScript.
         * @returns an "error" chunk if one is detected, otherwise whatever `brightScriptChunk` returns
         */
        function hashError() {
            if (check(lexer_1.Lexeme.HashError)) {
                let hashError = advance();
                let message = advance();
                return new CC.Error(hashError, message.text || "");
            }
            return brightScriptChunk();
        }
        /**
         * Parses tokens to produce a chunk of BrightScript.
         * @returns a chunk of plain BrightScript if any is detected, otherwise `undefined` to indicate
         *          that no non-conditional compilation directives were found.
         */
        function brightScriptChunk() {
            let chunkTokens = [];
            while (!check(lexer_1.Lexeme.HashIf, lexer_1.Lexeme.HashElseIf, lexer_1.Lexeme.HashElse, lexer_1.Lexeme.HashEndIf, lexer_1.Lexeme.HashConst, lexer_1.Lexeme.HashError)) {
                chunkTokens.push(advance());
                if (isAtEnd()) {
                    break;
                }
            }
            if (chunkTokens.length > 0) {
                return new CC.BrightScript(chunkTokens);
            }
            else {
                return undefined;
            }
        }
        function eof() {
            if (isAtEnd()) {
                return new CC.BrightScript([peek()]);
            }
            else {
                return undefined;
            }
        }
        function match(...lexemes) {
            for (let lexeme of lexemes) {
                if (check(lexeme)) {
                    advance();
                    return true;
                }
            }
            return false;
        }
        function consume(message, ...lexemes) {
            let foundLexeme = lexemes
                .map(lexeme => peek().kind === lexeme)
                .reduce((foundAny, foundCurrent) => foundAny || foundCurrent, false);
            if (foundLexeme) {
                return advance();
            }
            return emitError(new parser_1.ParseError(peek(), message));
        }
        function advance() {
            if (!isAtEnd()) {
                current++;
            }
            return previous();
        }
        function check(...lexemes) {
            if (isAtEnd()) {
                return false;
            }
            return lexemes.some(lexeme => peek().kind === lexeme);
        }
        function isAtEnd() {
            return peek().kind === lexer_1.Lexeme.Eof;
        }
        function peek() {
            return tokens[current];
        }
        function previous() {
            return tokens[current - 1];
        }
    }
}
exports.Parser = Parser;
